### PR TITLE
Multi-cluster k8s forno

### PR DESCRIPTION
**Pull request by tkporter**
### Description

This PR builds upon #4821, that's part of the reason why the diff is so huge!

#### Overview

This PR allows devs to use celotool to easily deploy full nodes in different clusters for the same celo env by allowing each celo env to have multiple contexts. For example, `rc1` could be the celo env and it could have contexts `gcp-us-west` and `gcp-asia-northeast`. It also introduces a new Terraform template for deploying a forno load balancer that routes traffic to full nodes in specified contexts.

#### Contexts

Contexts names are declared in the `.env` file along with their relevant configurations.
I think examples are the best way of showing what is going on:
```
# names of all supported contexts. 2 GCP context, one azure context, and 1 aws context
CONTEXTS=gcp-us-west,gcp-asia-northeast,azure-eastus,aws-us-west

# forno for this env will point to both GCP contexts
FORNO_CONTEXT=gcp-us-west1-a,gcp-asia-northeast2-a
FORNO_DOMAINS=this-is-a-test-forno.celo-networks-dev.org.
# all clusters must be in this VPC
FORNO_VPC_NETWORK_NAME=default

# definition of gcp-us-west ------------
# name of the GCP project
GCP_US_WEST_GCP_PROJECT_NAME=celo-testnet
# name of the GCP zone for the context
GCP_US_WEST_GCP_ZONE=us-west4-a
# name of the k8s cluster for the context
GCP_US_WEST_KUBERNETES_CLUSTER_NAME=federated-dev-us-west4-a
# how many full nodes in this context
GCP_US_WEST_FULL_NODES_COUNT=1
# the disk size of the full nodes in this context
GCP_US_WEST_FULL_NODES_DISK_SIZE=10

# definition of gcp-asia-northeast ------------
# name of the GCP project
GCP_ASIA_NORTHEAST_GCP_PROJECT_NAME=celo-testnet
# name of the GCP zone for the context
GCP_ASIA_NORTHEAST_GCP_ZONE=asia-northeast2-a
# name of the k8s cluster for the context
GCP_ASIA_NORTHEAST_KUBERNETES_CLUSTER_NAME=federated-dev-asia-northeast2-a
# how many full nodes in this context
GCP_ASIA_NORTHEAST_FULL_NODES_COUNT=1
# the disk size of the full nodes in this context
GCP_ASIA_NORTHEAST_FULL_NODES_DISK_SIZE=10

# definition of azure-eastus ----------------
# Azure subscription ID
AZURE_EASTUS_AZURE_SUBSCRIPTION_ID=97e2b592-255b-4f92-bce0-127257163c36
# Azure tenant ID
AZURE_EASTUS_AZURE_TENANT_ID=7cb7628a-e37c-4afb-8332-2029e418980e
# name of the resource group where the cluster lives
AZURE_EASTUS_AZURE_KUBERNETES_RESOURCE_GROUP=testnet-oracle-eastus
# name of the cluster
AZURE_EASTUS_KUBERNETES_CLUSTER_NAME=testnet-oracle-eastus
# some oracle specific things
AZURE_EASTUS_ORACLE_ADDRESS_AZURE_KEY_VAULTS=0x21860ca3a0a6f7e450b8f24bd00eac7ba766b85e:testnet-oracle-eastus,0xfd3738e5e0a020614a9e5253078dda491e77031c:testnet-oracle-eastus
# how many full nodes
AZURE_EASTUS_FULL_NODES_COUNT=2
# disk size for the full nodes
AZURE_EASTUS_FULL_NODES_DISK_SIZE=10

# definition of aws-us-west -----------------------
# name of the region where the k8s cluster is
AWS_US_WEST_AWS_KUBERNETES_CLUSTER_REGION=us-west-2
# name of the k8s cluster
AWS_US_WEST_KUBERNETES_CLUSTER_NAME=adorable-monster-1597251246
# how many full nodes
AWS_US_WEST_FULL_NODES_COUNT=2
AWS_US_WEST_FULL_NODES_DISK_SIZE=10
```

One thing to note is that the cluster's cloud platform is derived from its prefix, ie AWS contexts must begin with `aws-`, GCP `gcp-`, and AZURE `azure-`. Contexts have a "readable" form that can include lowercase letters and dashes, but ultimately behind the scenes this will be converted to all caps and underscores. When interacting with a celotool command that supports contexts (like `fullnodes`, `forno`, or `oracle`) you can pass the readable or all-caps version to `--context`.

#### Full nodes

`celotool deploy (initial|upgrade|destroy) fullnodes` has been added, which will deploy full nodes for a particular context. It supports GCP, Azure, and AWS contexts. If `--staticNodes` is provided, the full nodes will have nodekeys that are generated deterministically using the env's mnemonic and a hash of the name of the context. It will also upload the enodes in the deployment to the `static_nodes` GCP bucket with the name `<env>.<context>`. The `oracle-fullnode` deployment commands have been removed in favor of the generic `fullnodes` command.

#### Forno

`celotool deploy (initial|upgrade|destroy) forno` will deploy the `forno` terraform module to use contexts `FORNO_CONTEXT` as backends. It's basically a Terraform-ified version of the approach in https://blog.doit-intl.com/how-to-setup-multi-cluster-load-balancing-with-gke-4b407e1f3dff. It takes advantage of a cool feature GKE has called [container-native load balancing](https://cloud.google.com/kubernetes-engine/docs/how-to/container-native-load-balancing). It allows you to set a special annotation on a k8s service, and GKE will create a corresponding Network Endpoint Group (aka NEG) that can be used for GCE-based load balancing. The terraform module will then create resources to use [GCE's External HTTPS load balancing](https://cloud.google.com/load-balancing/docs/https) and send traffic to the cluster NEGs. The load balancer has a single global IP address that receives all ingress traffic, and a DNS A record must be created for each of the `FORNO_DOMAINS` that points to that IP. A GCP managed SSL certificate is also created by the Terraform module. All the heavy lifting of figuring out where to send traffic is done by GCP: https://cloud.google.com/load-balancing/docs/https#cross-region_load_balancing

Ultimately, you'll be able to send HTTP and WS RPC traffic to any of the `FORNO_DOMAINS`:
```
# http
https://mydomain.org

# ws
wss://mydomain.org/ws
```

#### other

I still need to add some more explanations into the README on the env variables to try to avoid confusion, and I'll also be creating a runbook.

### Tested

Lots of running :)

### Related issues

- Fixes #[issue number here]

### Backwards compatibility

_Brief explanation of why these changes are/are not backwards compatible._

### CODE REVIEW

File: packages/celotool/src/cmds/deploy/initial/fullnodes.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+type FullNodeInitialArgv = InitialArgv &
+  ContextArgv & {
+    createNeg: boolean
+    staticNodes: boolean
+  }
+
+export const builder = (argv: yargs.Argv) => {
+  return addContextMiddleware(argv)
+    .option('createNeg', {
```
Comment (trianglesphere) [inline packages/celotool/src/cmds/deploy/initial/fullnodes.ts]: Nit: `network endpoint group` -> `NEG` as an acronym

File: packages/helm-charts/celo-fullnode/templates/statefulset.yaml [hunk 2]
```diff
@@ -23,20 +24,64 @@ spec:
...
       labels:
-{{ include "celo-fullnode.labels" .  | indent 8 }}
+{{ include "common.standard.labels" .  | indent 8 }}
+        component: celo-fullnode
 {{ if .Values.prometheus | default true }}
       annotations:
 {{ include "common.prometheus-annotations" . | indent 8 }}
 {{ end }}
     spec:
+      # see https://www.verygoodsecurity.com/blog/posts/kubernetes-multi-az-deployments-using-pod-anti-affinity
```
Comment (trianglesphere) [inline packages/helm-charts/celo-fullnode/templates/statefulset.yaml:39]: Nit: This link appears to have already died (though a blurb does still exist on their blog feed)

File: packages/celotool/src/lib/forno.ts [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+import { coerceContext, getClusterManagerForContext, readableContext } from './context-utils'
+import { envVar, fetchEnv } from './env-utils'
+import { GCPClusterConfig } from './k8s-cluster/gcp'
+import { TerraformVars } from './terraform'
+import { deployModule, destroyModule } from './vm-testnet-utils'
+
+const FORNO_TERRAFORM_MODULE_NAME = 'forno'
+
+export async function deployForno(celoEnv: string) {
+  const contexts: string[] = fetchEnv(envVar.FORNO_FULL_NODE_CONTEXTS).split(',').map(coerceContext)
```
Comment (trianglesphere) [inline packages/celotool/src/lib/forno.ts:12]: My understanding is that gcp is the only valid forno full node context, but I never saw that written anywhere. If so (or if not) it'd be nice to have that explicitly written

**Review by trianglesphere — COMMENTED** [review]
I'm not that familiar with infra, but nothing stuck out to me as a large problem. +1 for the docs. They're super well written and gave me a good overview of what's going on.

File: packages/helm-charts/celo-fullnode/templates/statefulset.yaml (hunk 2 above)
Comment (tkporter) [reply packages/helm-charts/celo-fullnode/templates/statefulset.yaml:39]: Hmm weird, it works for me

Comment (trianglesphere) [reply packages/helm-charts/celo-fullnode/templates/statefulset.yaml:39]: Nvm, I checked again and it's working now.

File: packages/celotool/src/lib/prometheus.ts [hunk 4]
```diff
@@ -59,7 +59,38 @@ export async function upgradePrometheus() {
   return upgradeGenericHelmChart(kubeNamespace, releaseName, helmChartPath, await helmParameters())
 }
 
-async function helmParameters(clusterConfig?: AzureClusterConfig) {
+async function helmParameters(clusterConfig?: BaseClusterConfig) {
+  // To save $, don't send metrics to SD that probably won't be used
```
Comment (jcortejoso) [inline packages/celotool/src/lib/prometheus.ts:63]: Sweeet

**Review by jcortejoso — APPROVED** [review]
awesome work. I did a fast review of the changes and looks pretty well.

**Pull request merged**